### PR TITLE
bugfix: fixed template validation panic in case of incorrect ChangeScript configuration

### LINKED ISSUE #14367: Nomad job plan errors with the below logs.
### Nomad version
1.3.4

### Operating system and Environment details
Debian ARM64

### Issue

When doing a nomad job plan, I am getting the following error
`Error submitting job: Put "http://100.66.210.90:4646/v1/jobs?namespace=default&region=home": EOF`
The plan was working if change_mode is NOT set to script.

### Reproduction steps
Added `change_mode="script"` and `change_script` to the job template stanza.

#### Expected Result
Plan works as expected.

#### Actual Result
Error submitting job: Put "http://100.66.210.90:4646/v1/jobs?namespace=default&region=home": EOF

### Job file (if appropriate)

Here is the snippet

```
          "Templates": [
            {
              "SourcePath": "",
              "DestPath": "local/Caddyfile",
              "EmbeddedTmpl": "<REDACTED>",
              "ChangeMode": "script",
              "ChangeSignal": "",
              "ChangeScript": null,
              "Splay": 5000000000,
              "Perms": "0644",
              "Uid": null,
              "Gid": null,
              "LeftDelim": "[[",
              "RightDelim": "]]",
              "Envvars": false,
              "VaultGrace": 0,
              "Wait": null
            }
          ],
```

### Nomad Client logs (if appropriate)

```
Aug 28 19:24:36 rpi0 nomad[4032]:     2022-08-28T19:24:36.784+0530 [ERROR] http: http: panic serving 100.111.170.15:53910: runtime error: invalid memor>
Aug 28 19:24:36 rpi0 nomad[4032]: goroutine 2003 [running]:
Aug 28 19:24:36 rpi0 nomad[4032]: net/http.(*conn).serve.func1()
Aug 28 19:24:36 rpi0 nomad[4032]:         net/http/server.go:1850 +0xb8
Aug 28 19:24:36 rpi0 nomad[4032]: panic({0x2161220, 0x3e68400})
Aug 28 19:24:36 rpi0 nomad[4032]:         runtime/panic.go:890 +0x260
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Template).Validate(0x4000794600)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:7825 +0x30c
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Task).Validate(0x40017ee6e0, 0x400195a1b0, {0x400220e0d6, 0x6}, {0x400283c>
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:7297 +0x13c8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*TaskGroup).Validate(0x4000bc07e0, 0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:6517 +0x21a0
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Job).Validate(0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:4446 +0x1cdc
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.jobValidate.Validate({}, 0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:248 +0x38
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).admissionValidators(0x400083cc30, 0x46daf0?)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:104 +0xf8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).admissionControllers(0x40001161c0?, 0x2673a55?)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:71 +0x44
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).Plan(0x400083cc30, 0x4001182a80, 0x4000783ce0)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint.go:1606 +0xc4
Aug 28 19:24:36 rpi0 nomad[4032]: reflect.Value.call({0x4000862900?, 0x40001105c0?, 0x4e40?}, {0x266bf83, 0x4}, {0x40022493e8, 0x3, 0x4000783ce0?})
Aug 28 19:24:36 rpi0 nomad[4032]:         reflect/value.go:584 +0x688
Aug 28 19:24:36 rpi0 nomad[4032]: reflect.Value.Call({0x4000862900?, 0x40001105c0?, 0x251ea00?}, {0x40022493e8?, 0x251ea00?, 0x22a7c20?})
Aug 28 19:24:36 rpi0 nomad[4032]:         reflect/value.go:368 +0x90
Aug 28 19:24:36 rpi0 nomad[4032]: net/rpc.(*service).call(0x40001163c0, 0x2c3bd28?, 0x40022202c0?, 0x0, 0x40000b9e00, 0x257d540?, {0x251ea00?, 0x400118>
Aug 28 19:24:36 rpi0 nomad[4032]:         net/rpc/server.go:382 +0x20c
Aug 28 19:24:36 rpi0 nomad[4032]: net/rpc.(*Server).ServeRequest(0x4000f31578?, {0x2c3bd28, 0x40022202c0})
Aug 28 19:24:36 rpi0 nomad[4032]:         net/rpc/server.go:503 +0x110
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Server).RPC(0x400099c380, {0x2673a55, 0x8}, {0x251ea00?, 0x4001182900}, {0x22a7c20>
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/server.go:1891 +0xf8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/command/agent.(*Agent).RPC(0x40011bf1d0?, {0x2673a55?, 0x4000cf6000?}, {0x251ea00?, 0x4001>
```

### PR DESCRIPTION
**Pull request by pkazmierczak**
Fixes #14367

### CODE REVIEW

File: nomad/structs/structs.go [hunk 1]
```diff
@@ -7739,6 +7725,10 @@ func (t *Template) Copy() *Template {
 		nt.Wait = t.Wait.Copy()
 	}
 
+	if t.ChangeScript != nil {
+		nt.ChangeScript = t.ChangeScript.Copy()
+	}
```
Comment (tgross) [inline nomad/structs/structs.go:7730]: We nil check the receiver in the `Copy` and return nil, so we can safely leave it out here.

File: nomad/structs/structs.go [hunk 2]
```diff
@@ -7822,6 +7816,44 @@ func (t *Template) DiffID() string {
...
+func (cs *ChangeScript) Copy() *ChangeScript {
+	if cs == nil {
+		return nil
+	}
+
+	ncs := new(ChangeScript)
+	*ncs = *cs
+
+	return ncs
+}
```
Comment (tgross) [inline nomad/structs/structs.go:7842]: Unfortunately you'll need to copy the `Args` field as well. Fortunately we have [`helper.CopySlice`](https://github.com/hashicorp/nomad/blob/main/helper/funcs.go#L289-L301). See https://play.golang.com/p/JAnuZGsPlBw

**Review by tgross — CHANGES_REQUESTED** [review]
Thanks for the quick fix @pkazmierczak!

File: nomad/structs/structs.go (hunk 2 above)
Comment (pkazmierczak) [reply nomad/structs/structs.go:7842]: uff, of course, thanks for catching this. fixed.

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -7822,6 +7812,47 @@ func (t *Template) DiffID() string {
...
+func (cs *ChangeScript) Copy() *ChangeScript {
+	if cs == nil {
+		return nil
+	}
+
+	ncs := new(ChangeScript)
+	*ncs = *cs
+
+	// args is a slice!
+	ncs.Args = slices.Clone(cs.Args)
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply nomad/structs/structs.go:7838]: yep. stumbled upon the same comment.

**Pull request merged**